File: SW1/python/1/src/contact_manager/commands.py
from __future__ import annotations

from abc import ABC, abstractmethod

from contact_manager.models import Contact
from contact_manager.repository import SQLiteContactRepository


class Command(ABC):
    """Command interface for action execution and undo."""

    @abstractmethod
    def execute(self) -> Contact | None:
        raise NotImplementedError

    @abstractmethod
    def undo(self) -> None:
        raise NotImplementedError


class AddContactCommand(Command):
    def __init__(self, repository: SQLiteContactRepository, contact: Contact) -> None:
        self._repository = repository
        self._contact = contact
        self._created_contact: Contact | None = None

    def execute(self) -> Contact:
        self._created_contact = self._repository.add(self._contact)
        return self._created_contact

    def undo(self) -> None:
        if self._created_contact is not None and self._created_contact.contact_id is not None:
            self._repository.delete(self._created_contact.contact_id)


class UpdateContactCommand(Command):
    def __init__(self, repository: SQLiteContactRepository, updated_contact: Contact) -> None:
        self._repository = repository
        self._updated_contact = updated_contact
        self._original_contact: Contact | None = None

    def execute(self) -> Contact:
        if self._updated_contact.contact_id is None:
            raise ValueError("Cannot update a contact without an id.")

        self._original_contact = self._repository.get(self._updated_contact.contact_id)
        if self._original_contact is None:
            raise ValueError("Contact not found.")

        self._repository.update(self._updated_contact)
        return self._updated_contact

    def undo(self) -> None:
        if self._original_contact is not None:
            self._repository.update(self._original_contact)


class DeleteContactCommand(Command):
    def __init__(self, repository: SQLiteContactRepository, contact_id: int) -> None:
        self._repository = repository
        self._contact_id = contact_id
        self._deleted_contact: Contact | None = None

    def execute(self) -> Contact | None:
        self._deleted_contact = self._repository.delete(self._contact_id)
        return self._deleted_contact

    def undo(self) -> None:
        if self._deleted_contact is not None:
            restored = self._repository.add(self._deleted_contact.clone(contact_id=None))
            self._deleted_contact = restored


class CommandManager:
    """Stores command history and supports undo."""

    def __init__(self) -> None:
        self._history: list[Command] = []

    def execute(self, command: Command) -> Contact | None:
        result = command.execute()
        self._history.append(command)
        return result

    def undo_last(self) -> bool:
        if not self._history:
            return False

        command = self._history.pop()
        command.undo()
        return True

File: SW1/csharp/1/Commands.cs
namespace ContactManager;

public interface ICommand
{
    Contact? Execute();
    void Undo();
}

public sealed class AddContactCommand : ICommand
{
    private readonly SQLiteContactRepository _repository;
    private readonly Contact _contact;
    private Contact? _createdContact;

    public AddContactCommand(SQLiteContactRepository repository, Contact contact)
    {
        _repository = repository;
        _contact = contact;
    }

    public Contact Execute()
    {
        _createdContact = _repository.Add(_contact);
        return _createdContact;
    }

    public void Undo()
    {
        if (_createdContact?.ContactId is int contactId)
        {
            _repository.Delete(contactId);
        }
    }
}

public sealed class UpdateContactCommand : ICommand
{
    private readonly SQLiteContactRepository _repository;
    private readonly Contact _updatedContact;
    private Contact? _originalContact;

    public UpdateContactCommand(SQLiteContactRepository repository, Contact updatedContact)
    {
        _repository = repository;
        _updatedContact = updatedContact;
    }

    public Contact Execute()
    {
        if (_updatedContact.ContactId is null)
        {
            throw new InvalidOperationException("Cannot update a contact without an id.");
        }

        _originalContact = _repository.Get(_updatedContact.ContactId.Value)
            ?? throw new InvalidOperationException("Contact not found.");

        _repository.Update(_updatedContact);
        return _updatedContact;
    }

    public void Undo()
    {
        if (_originalContact is not null)
        {
            _repository.Update(_originalContact);
        }
    }
}

public sealed class DeleteContactCommand : ICommand
{
    private readonly SQLiteContactRepository _repository;
    private readonly int _contactId;
    private Contact? _deletedContact;

    public DeleteContactCommand(SQLiteContactRepository repository, int contactId)
    {
        _repository = repository;
        _contactId = contactId;
    }

    public Contact? Execute()
    {
        _deletedContact = _repository.Delete(_contactId);
        return _deletedContact;
    }

    public void Undo()
    {
        if (_deletedContact is not null)
        {
            _deletedContact = _repository.Add(
                _deletedContact.Clone(contactId: null, keepExistingId: false)
            );
        }
    }
}

public sealed class CommandManager
{
    private readonly Stack<ICommand> _history = new();

    public Contact? Execute(ICommand command)
    {
        var result = command.Execute();
        _history.Push(command);
        return result;
    }

    public bool UndoLast()
    {
        if (_history.Count == 0)
        {
            return false;
        }

        _history.Pop().Undo();
        return true;
    }
}
